File: 7-kyu/binary-tree-search-(not-bst)/c++/solution.cpp
struct Node {
  int val;
  Node *left = nullptr;
  Node *right = nullptr;
};

bool search(int n, Node *root)
{
    if (root == nullptr) return false;
    else if (root->val == n) return true;
    else {
        if (root->left == nullptr && root->right == nullptr && n != root->val) return false;
        return search(n, root->left) || search(n, root->right);
    }
}

File: 7-kyu/binary-tree-search-(not-bst)/python/solution.py
from __future__ import annotations
from typing import Optional


class Node:
    def __init__(self, value: int, left: Optional[Node] = None, right: Optional[Node] = None):
        self.value = value
        self.left = left
        self.right = right
        
    
def search(n: int, root: Optional[Node]) -> bool:
    """ Determines if a value is in a binary tree (NOT bst) """
    if root == None:
        return False
    elif root.value == n:
        return True
    elif root.left == None and root.right == None and root.value != n:
        return False
    return search(n, root.left) or search(n, root.right)
